packages/react-core/src/components/TimePicker/TimePicker.tsx

```diff
@@ -0,0 +1,457 @@
+    if (this.inputRef && this.inputRef.current && this.inputRef.current.contains(event.target as Node)) {
+      if (!isOpen && event.key !== KeyTypes.Tab) {
+        this.onToggle(true);
+      } else if (isOpen) {
+        if (event.key === KeyTypes.Escape) {
+          this.onToggle(false);
+        } else if (event.key === KeyTypes.Tab) {
+          this.onToggle(false);
+        } else if (event.key === KeyTypes.Enter) {
+          if (focusedIndex !== null) {
+            this.onSelect((this.refCollection[focusedIndex] as HTMLElement).innerText, focusedIndex);
+            event.stopPropagation();
+          } else {
+            this.onToggle(false);
+          }
+        } else if (event.key === KeyTypes.ArrowDown) {
+          this.updateFocusedIndex(1);
+          event.preventDefault();
+        } else if (event.key === KeyTypes.ArrowUp) {
+          this.updateFocusedIndex(-1);
+          event.preventDefault();
+        }
+      }
+    }
+  };
+
+  componentDidUpdate(prevProps: TimePickerProps, prevState: TimePickerState) {
+    const { time } = this.state;
+    if (this.state.isOpen && !prevState.isOpen && this.state.time && !this.state.isInvalid) {
+      this.scrollToSelection(time);
+    }
+  }
+
+  updateFocusedIndex = (increment: number) => {
+    this.setState(prevState => {
+      const maxIndex = this.refCollection.length - 1;
+      let nextIndex = prevState.focusedIndex !== null ? prevState.focusedIndex + increment : prevState.scrollIndex;
+      if (nextIndex < 0) {
+        nextIndex = maxIndex;
+      } else if (nextIndex > maxIndex) {
+        nextIndex = 0;
+      }
+      this.scrollToIndex(nextIndex);
+      return {
+        focusedIndex: nextIndex,
+        time: this.refCollection[nextIndex].innerText
+      };
+    });
+  };
+
+  scrollToIndex = (index: number) => {
+    this.refCollection[index].offsetParent.scrollTop = this.refCollection[index].offsetTop;
+  };
+
+  scrollToSelection = (time: string) => {
+    const { delimiter, variant } = this.props;
+    let splitTime = time.split(this.props.delimiter);
+    let focusedIndex = null;
+
+    // build out the rest of the time assuming hh:00 if it's a partial time
+    if (splitTime.length < 2) {
+      time = `${time}${delimiter}00`;
+      splitTime = time.split(delimiter);
+    }
+
+    // for 12hr variant, autoscroll to pm if it's currently the afternoon, otherwise autoscroll to am
+    if (variant === '12hr' && splitTime.length > 1 && splitTime[1].length < 2) {
+      const minutes = splitTime[1].length === 0 ? '00' : splitTime[1] + '0';
+      time = `${splitTime[0]}${delimiter}${minutes}${new Date().getHours() > 11 ? 'pm' : 'am'}`;
+    } else if (
+      variant === '12hr' &&
+      splitTime.length > 1 &&
+      splitTime[1].length === 2 &&
+      !time.toLowerCase().includes('am') &&
+      !time.toLowerCase().includes('pm')
+    ) {
+      time = `${time}${new Date().getHours() > 11 ? 'pm' : 'am'}`;
+    }
```

Comment:
I'll pull out the code they share into a new `appendMissingAmPm` function.

Reply:
I lied - they don't have any shared code/logic. It looks similar, probably because it's betraying something about the way my brain works, but there is enough difference that they actually dont share any scenarios.

`parseTime` parses ISO 8601 date strings, and appends am/pm to 12 hour times that are otherwise valid.

`scrollToSection` finds the option that matches the current partial or complete time input and scrolls to it. In the case that there is not exact match, it determines the closest match and scrolls to it.

Reply:
That's fine. I wonder if we could embed each option's time in a data attribute in the `<li>` to avoid this in the future.

Reply:
how would that change this situation? it would just be the same as the option's value?

Reply:
I meant that you could put the Date object into a data attribute and then compare the time part to what the user has input.

`<li data-date="2020-12-10T15:00:00.000Z ... />` -> `new Date(ref.getAttribute('data-date'))`

This way entering `1:5` could snap to `02:00` instead of `01:00` like it does now.